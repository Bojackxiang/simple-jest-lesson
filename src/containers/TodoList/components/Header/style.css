*{
    margin: 0;
    padding: 0; 
}

.header {
    /* line-height:60px; */
    background-color: #333;
    font-size:24px; 
    color: #fff; 
}

.header-content {
    width: 600px; 
    margin: 0 auto;
    display: flex;
    justify-content: space-between;
}