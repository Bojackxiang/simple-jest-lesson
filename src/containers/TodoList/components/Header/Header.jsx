import React, { Component } from "react";
import { connect } from "react-redux";
import { changeInputValue } from "../../store/actions";

import "./style.css";
class Header extends Component {
  constructor(props) {
    super(props);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.passValue = this.passValue.bind(this);

    this.state = {
      value: "",
    };
  }

  handleInputChange(e) {
    this.setState({
      ...this.state,
      value: e.target.value,
    });
  }

  handleKeyUp(e) {
    const { value } = this.state;
    if (e.keyCode === 13 && this.state.value) {
      this.props.updateList(value);
      this.setState({
        ...this.state,
        value: "",
      });
    }
  }

  passValue(value) {
    console.log(value);
  }

  render() {
    return (
      <div className="header">
        <div className="header-content">
          <h3>Todo</h3>
          <input
            data-test="input"
            value={this.props.value}
            onChange={(e) => this.props.handleInputChange(e.target.value)}
            onKeyUp={this.handleKeyUp}
          />
        </div>
      </div>
    );
  }
}

const mapState = (state) => {
  return {
    value: state.todo.inputValue,
  };
};

const mapDispatch = (dispatch) => ({
  handleInputChange(value) {
    dispatch(changeInputValue(value));
  },
});

export default connect(mapState, mapDispatch)(Header);
