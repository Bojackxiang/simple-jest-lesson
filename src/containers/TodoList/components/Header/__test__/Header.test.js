import { shallow } from "enzyme";
import Header from "../Header";

test("[Header] should have an input element", () => {
  const wrapper = shallow(<Header />);
  const input = wrapper.find(dataTestName("input"));

  expect(input.length).toBe(1);
});

test("[Header] should have empty value when init", () => {
  const wrapper = shallow(<Header />);
  const input = wrapper.find(dataTestName("input"));
  const initValue = input.prop("value");

  expect(initValue).toEqual("");
});

test("[Header] state and input value should updated", () => {
  const wrapper = shallow(<Header />);
  const input = wrapper.find(dataTestName("input"));
  const updatedValue = "updated";
  // simulate 的使用
  input.simulate("change", {
    target: {
      value: updatedValue,
    },
  });
  // 这边需要小心，这里的state 是 wrapper 的state
  expect(wrapper.state("value")).toEqual(updatedValue);
  //  ! input 老的element 还是原来的 值
  const newInput = wrapper.find(dataTestName("input"));
  expect(newInput.prop("value")).toBe(updatedValue);
});

test("[Header] when press enter, if not value should do nothing", (done) => {
  const fn = jest.fn();
  const wrapper = shallow(<Header addUndoItem={fn} />);
  const input = wrapper.find(dataTestName("input"));
  const value = "";
  wrapper.setState({
    value: value,
  });

  // 回车
  input.simulate("keyUp", {
    keyCode: 13,
  });
  expect(fn).not.toHaveBeenCalled();
  done();
});

test("[Header] when press enter, fu should be triggered", (done) => {
  const fn = jest.fn();
  const wrapper = shallow(<Header addUndoItem={fn} />);
  const input = wrapper.find(dataTestName("input"));
  const value = "hello";
  wrapper.setState({
    value: value,
  });

  // 回车
  input.simulate("keyUp", {
    keyCode: 13,
  });
  expect(fn).toHaveBeenCalled();
  expect(fn).toHaveBeenCalledWith(value);
  done();
});

test("[Header] when press enter, the input should be cleared", (done) => {
  const fn = jest.fn();
  const wrapper = shallow(<Header addUndoItem={fn} />);
  const input = wrapper.find(dataTestName("input"));
  const value = "hello";
  wrapper.setState({
    value: value,
  });

  // 回车
  input.simulate("keyUp", {
    keyCode: 13,
  });

  const newInput = wrapper.find(dataTestName("input"));
  const newInputValue = newInput.prop("value");
  expect(newInputValue).toBe("");

  done();
});

export const dataTestName = (name) => `[data-test="${name}"]`;
