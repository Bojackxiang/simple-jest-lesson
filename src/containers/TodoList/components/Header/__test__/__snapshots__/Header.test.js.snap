// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Header] snapshot test 1`] = `
<div
  className="header"
>
  <h3>
    Todo
  </h3>
  <input
    data-test="input"
    onChange={[Function]}
    onKeyUp={[Function]}
    value=""
  />
</div>
`;
